### ISSUE TITLE
UIMenuItem: isDisabled() is not checked inside decode()

### ISSUE THREAD

**Issue by mmariotti** [primefaces/primefaces#5597]
## 1) Environment
- PrimeFaces version: **ALL**
- Does it work on the newest released PrimeFaces version? Version? **NO**
- Does it work on the newest sources in GitHub? **NO**
- Application server + version: **ALL**
- Affected browsers: **ALL**

## 2) Expected behavior
`UIMenuItem` should conform to every other `UICommand` component, checking for `isDisabled()` before enqueuing an `ActionEvent` inside `decode()` method.

## 3) Actual behavior
It is possible to invoke a **DISABLED** `UIMenuItem` using javascript, leading to execution of unintended code on server

## 4) Steps to reproduce
create a 
```xml
<h:form id="form">
...
<p:menuitem id="menuitem" disabled="true" action="#{someBean.someAction}" value="disabled menuitem" />
...
</h:form>
```
and execute `PrimeFaces.ab({s:"form:menuitem",p:"form",u:"form",f:"form"})`.
Despite the **disabled="true"**, the menuitem is **decoded** and the action is invoked!

## 5) Sample XHTML
```xml
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns:ui="http://xmlns.jcp.org/jsf/facelets"
  xmlns:h="http://xmlns.jcp.org/jsf/html" xmlns:f="http://xmlns.jcp.org/jsf/core" xmlns:p="http://primefaces.org/ui">

<h:head>
</h:head>

<h:body>
  <h:form id="form">
    <p:panel>
      <ul>
        <li>
          execute
          <pre>PrimeFaces.ab({s:"form:commandButton2",p:"form",u:"form",f:"form"})</pre>
          and the following timestamp will be resetted to "[NOT SET]".
          <br />
          This means that "#{'#{testMenuitemBean.updateTimestamp}'}" has not been invoked (testMenuitemBean is
          @RequestScoped).
        </li>
        <br />
        <br />
        <li>
          execute
          <pre>PrimeFaces.ab({s:"form:menuitem2",p:"form",u:"form",f:"form"})</pre>
          and the following timestamp
          <h1>will be updated!!</h1>
        </li>
      </ul>
    </p:panel>

    <br />

    <p:panel id="timestampPanel">timestamp: [#{testMenuitemBean.timestamp}]</p:panel>

    <br />

    <p:panel>
      <p:menuButton id="menuButton1" value="menuButton1">
        <p:menuitem id="menuitem1" action="#{testMenuitemBean.updateTimestamp}" process="@form" update="timestampPanel"
          value="enabled menuitem" />
        <p:menuitem id="menuitem2" disabled="true" action="#{testMenuitemBean.updateTimestamp}" process="@form"
          update="timestampPanel" value="disabled menuitem" />
      </p:menuButton>

      <p:commandButton id="commandButton1" action="#{testMenuitemBean.updateTimestamp}" process="@form"
        update="timestampPanel" value="enabled commandButton" />
      <p:commandButton id="commandButton2" disabled="true" action="#{testMenuitemBean.updateTimestamp}" process="@form"
        update="timestampPanel" value="disabled commandButton" />
    </p:panel>
  </h:form>
</h:body>
</html>
```

## 6) Sample bean
```java
import javax.enterprise.context.RequestScoped;
import javax.inject.Named;

@Named
@RequestScoped
public class TestMenuitemBean {
  private String timestamp = "NOT SET";

  public void updateTimestamp() {
    timestamp = String.valueOf(System.currentTimeMillis());
  }

  public String getTimestamp() {
    return timestamp;
  }
}
```

## 7) Solution
this is the current `decode()` of `https://github.com/primefaces/primefaces/blob/master/src/main/java/org/primefaces/component/menuitem/UIMenuItem.java`
```java
@Override
public void decode(FacesContext facesContext) {
  Map<String, String> params = facesContext.getExternalContext().getRequestParameterMap();
  String clientId = getClientId(facesContext);

  if (params.containsKey(clientId)) {
    queueEvent(new ActionEvent(this));
  }

  ComponentUtils.decodeBehaviors(facesContext, this);
}
```

just add:

```java
@Override
public void decode(FacesContext facesContext) {
  if(isDisabled()) {             // <------------   this check
    return;
  }

  Map<String, String> params = facesContext.getExternalContext().getRequestParameterMap();
  String clientId = getClientId(facesContext);

  if (params.containsKey(clientId)) {
    queueEvent(new ActionEvent(this));
  }

  ComponentUtils.decodeBehaviors(facesContext, this);
}
```

However, since `UIMenuItem` has to be used inside another "menu-enabled" component (like `TieredMenu` , `SlideMenu`, `MenuButton`, ...), it would be better to check also if the "menu-enabled" ancestor is disabled.
As of now, I didn't found a simpler way to reach that ancestor, so I'm doing:

```java
@Override
public void decode(FacesContext facesContext)
{
  if(isDisabled()) {
    return;
  }

  UIComponent ancestor = getParent(); 
  while(ancestor instanceof UICommand || ancestor instanceof AbstractMenu || ancestor instanceof Submenu) {
    if(!ancestor.isRendered() || Boolean.valueOf(String.valueOf(ancestor.getAttributes().get("disabled")))) {
      return;
    }
    
    ancestor = ancestor.getParent();
  }

  super.decode(facesContext);
}
```

Thank you and good work!

**Comment by melloware** [comment]
Nice thoughtful issue!  @Rapster @tandraschko thoughts on this?

**Comment by mmariotti** [comment]
My application failed a PenTest because of this one... :(
I can't even soothe myself 'cause someone else spotted it :@

**Comment by Rapster** [comment]
> it would be better to check also if the "menu-enabled" ancestor is disabled.

Just so I can follow you, why that?

**Comment by mmariotti** [comment]
take the example and add:

```xml
<p:menuButton id="menuButton" value="menuButton2" disabled="true">
  <p:menuitem id="menuitem" action="#{testMenuitemBean.updateTimestamp}" process="@form" update="timestampPanel"
    value="enabled menuitem" />
</p:menuButton>
```
and now execute: `PrimeFaces.ab({s:"form:menuitem",p:"form",u:"form",f:"form"})`.

Even if the `<p:menuButton/>` is **disabled**, the inner `<p:menuitem/>` is not, and it's not even **rendered** (it's not present on the page source/dom but its attribute `rendered` is **true** and `disabled` is **false**).
So, if you execute the JS,  the **<p:menuitem/>** action is invoked anyway.
Since, generally, you don't put an `id` on every `<p:menuitem/>` it's easy to guess it's sequential clientId, like `form:j_idt893`.
Also, before this, i was confident that if the menuitem was inside a disabled "menu-container", and not rendered, it cannot be invoked. But it's not this way :/

**Comment by melloware** [comment]
@mmariotti Can you try this version in your local code.  I made some slight modifications.

1. I defensively check if the ancestor is not null. (not sure this can happen but better to be safe)

2. I use the Menu interfaces which I think captures all of your components and future ones/

```java
@Override
public void decode(FacesContext facesContext) {
     if (isDisabled()) {
         return;
     }

     UIComponent ancestor = getParent();
     while (ancestor != null && ancestor instanceof MenuElement || ancestor instanceof MenuItemAware) {
            if (!ancestor.isRendered() || Boolean.valueOf(String.valueOf(ancestor.getAttributes().get("disabled")))) {
                return;
            }
            ancestor = ancestor.getParent();
       }

     super.decode(facesContext);
}
```

**Comment by Rapster** [comment]
@mmariotti the behavior you're describing would happen with every containers/or components containing commands no? Is this really something we can prevent?

**Comment by mmariotti** [comment]
@melloware it's useless, it cannot happen because `null instanceof Anything` is always false.
@Rapster actually, it seems there's no container component that can be disabled while allowing `UICommand` children other than `AbstractMenu` descendants and `SplitButton`.
Nevertheless, it feels weird to check for somthing so specific, and does not take into account new future components.
But yes, it can be prevented, there's a solution that works globally.
It's not implemented on component side, but on application side, using a global `ActionListener`.
In this implementation, I'm looking for **any** invalid component in the ancestors of the command-component-to-be-invoked.
```java
import javax.faces.component.UIComponent;
import javax.faces.event.AbortProcessingException;
import javax.faces.event.ActionEvent;
import javax.faces.event.ActionListener;

import one.util.streamex.StreamEx;

public class SecurityActionListener implements ActionListener
{
    private ActionListener base;

    public SecurityActionListener(ActionListener base)
    {
        this.base = base;
    }

    @Override
    public void processAction(ActionEvent event) throws AbortProcessingException
    {
        UIComponent source = event.getComponent();
        if(source == null)
        {
            return;
        }

        boolean invalid = StreamEx.iterate(source, Objects::nonNull, UIComponent::getParent)
            .anyMatch(x -> Boolean.valueOf(String.valueOf(x.getAttributes().get("disabled"))) || Boolean.valueOf(String.valueOf(x.getAttributes().get("readonly"))));

        if(invalid)
        {
            throw new SecurityException("invocation is denied");
        }

        base.processAction(event);
    }
}
```
This has to be enabled in `faces-config.xml`:
```xml
<?xml version="1.0" encoding="utf-8"?>
<faces-config version="2.2" xmlns="http://xmlns.jcp.org/xml/ns/javaee" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-facesconfig_2_2.xsd">

    <application>
        <action-listener>it.shape.core.jsf.application.SecurityActionListener</action-listener>
        ...
        <action-listener>org.primefaces.application.DialogActionListener</action-listener>
        <navigation-handler>org.primefaces.application.DialogNavigationHandler</navigation-handler>
        <view-handler>org.primefaces.application.DialogViewHandler</view-handler>
        ...
    </application>
    ...
</faces-config>
```

Tomorrow I'll check if the same applies also to `ClientBehaviors`

**Comment by Rapster** [comment]
> it seems there's no container component that can be disabled while allowing UICommand

I think you're wrong, most of our component can `p:panel`,` p:dataTable`, `p:outputPanel` etc. using passthrough attributes at least

First, if we were to apply this fix (which looks correct) it's a performance killer. Depending the number of components in your tree and commands, this fix will have side effect

Second, in your sample let's say you disabled this button because the user is not allowed to. Even using your hack, the method called (service method probably) should be secured and throw an exception. For more details, see https://owasp-aasvs.readthedocs.io/en/latest/requirement-4.9.html

Finally, I think this fix could prevent some hacks yes and even though your impl `SecurityActionListener` looks better. I have one concern: performance. Is it worth to slow down performance for a use case that will happen rarely? And plus, as ASVS rule mentionned, security check should be done on server side (I think they mean business layer in our case)

Last thought, adding `SecurityActionListener` could be optional though it's up to the user to enable it or not, but again, IMO I think it's a workaround to the real problem

**Comment by mmariotti** [comment]
@Rapster 

> I think you're wrong, most of our component can `p:panel`,` p:dataTable`, `p:outputPanel` etc.

You are right, I focused on containers of just menuitems, but we should look also for generic containers that can have `disabled` and, eventually, `readonly` attributes.


> First, if we were to apply this fix (which looks correct) it's a performance killer. Depending the number of components in your tree and commands, this fix will have side effect

Yes, I know. See the final code below for a better compromise.


> Second, in your sample let's say you disabled this button because the user is not allowed to. Even using your hack, the method called (service method probably) should be secured and throw an exception. For more details, see https://owasp-aasvs.readthedocs.io/en/latest/requirement-4.9.html

> as ASVS rule mentionned, security check should be done on server side (I think they mean business layer in our case)

The definition does not specify the layer on purpose. It's necessary and sufficient condition that the same check is performed on server-side, any layer.
In this respect, the developer should be relieved of the burden and must be able to trust the framework as a delegate for server side access control rules enforcement, which is exactly the duty of JSF ViewState and Component-Tree.
In other words, you shouldn't be required to check twice by yourself (it's the JSF *killer-feature*), otherwise it's not JSF anymore, but another *spring-whatever-DIY-like* framework.
The framework is doing a great job at this, just need to fill some small hole.


> Finally, I think this fix could prevent some hacks yes and even though your impl SecurityActionListener looks better. I have one concern: performance. Is it worth to slow down performance for a use case that will happen rarely?

Checking the component hierarchy in `decode()` for every `UICommand` is overwhelming: it adds O(n * h) to the `APPLY_REQUEST_VALUES` phase.
Nevertheless, using a global `ActionListener` does not have a relevant impact on performance, since it's called **only** for the component to be invoked, and will just traverse the tree bottom-up, using O(h) complexity - which is negligible.
In the end, is it worth it? Of course.
Is there a better solution? There is, but it requires more work (see below).

> Last thought, adding SecurityActionListener could be optional though it's up to the user to enable it or not

I agree.

> IMO I think it's a workaround to the real problem

In this specific case, the **big** problem is that `UIMenuItem.decode()` does not check for `isDisabled()` in contrast with all others `UICommand` components.
More generally, allowing execution of a `UICommand` that is a descendant of a `disabled` component is not a theoretical violation, but it's clearly an **unwanted behavior**, since it's impossible to achive that behavior with *normal* user actions on the GUI, but it can be achieved with specially crafted JS.
This seems to me the exact definition of *vulnerability*.
 
---------------------------------------
 
This is the last version of the test page:
```xml
<h:form id="form">
    <p:panelGrid columns="3" style="width: 100%">
        <f:facet name="header">
            <p:outputPanel id="timestampPanel">timestamp: [#{testMenuitemBean.timestamp}]</p:outputPanel>
        </f:facet>

        <h:panelGroup>
            <p:commandButton id="b1" action="#{testMenuitemBean.updateTimestamp}" process="@form" update="timestampPanel"
                value="b1" />

            <div>
                <p:commandButton type="button" value="hack b1" onclick="PrimeFaces.ab({s:'form:b1',p:'form',u:'form',f:'form'})" />
            </div>
        </h:panelGroup>

        <h:panelGroup>
            <p:commandButton id="b2" disabled="true" action="#{testMenuitemBean.updateTimestamp}" process="@form"
                update="timestampPanel" value="b2" />

            <div>
                <p:commandButton type="button" value="hack b2" onclick="PrimeFaces.ab({s:'form:b2',p:'form',u:'form',f:'form'})" />
            </div>
        </h:panelGroup>


        <h:panelGroup>
            <p:commandButton id="b3" rendered="false" action="#{testMenuitemBean.updateTimestamp}" process="@form"
                update="timestampPanel" value="b3" />

            <div>
                <p:commandButton type="button" value="hack b3" onclick="PrimeFaces.ab({s:'form:b3',p:'form',u:'form',f:'form'})" />
            </div>
        </h:panelGroup>


        <h:panelGroup>
            <p:menuButton value="m1 - m2 - m3">
                <p:menuitem id="m1" action="#{testMenuitemBean.updateTimestamp}" process="@form" update="timestampPanel"
                    value="m1" />
                <p:menuitem id="m2" disabled="true" action="#{testMenuitemBean.updateTimestamp}" process="@form"
                    update="timestampPanel" value="m2" />
                <p:menuitem id="m3" rendered="false" action="#{testMenuitemBean.updateTimestamp}" process="@form"
                    update="timestampPanel" value="m3" />
            </p:menuButton>

            <div>
                <p:commandButton type="button" value="hack m1" onclick="PrimeFaces.ab({s:'form:m1',p:'form',u:'form',f:'form'})" />
                <p:commandButton type="button" value="hack m2" onclick="PrimeFaces.ab({s:'form:m2',p:'form',u:'form',f:'form'})"
                    style="background:salmon" />
                <p:commandButton type="button" value="hack m3" onclick="PrimeFaces.ab({s:'form:m3',p:'form',u:'form',f:'form'})" />
            </div>
        </h:panelGroup>


        <h:panelGroup>
            <p:menuButton value="m4 - m5 - m6" disabled="true">
                <p:menuitem id="m4" action="#{testMenuitemBean.updateTimestamp}" process="@form" update="timestampPanel"
                    value="m4" />
                <p:menuitem id="m5" disabled="true" action="#{testMenuitemBean.updateTimestamp}" process="@form"
                    update="timestampPanel" value="m5" />
                <p:menuitem id="m6" rendered="false" action="#{testMenuitemBean.updateTimestamp}" process="@form"
                    update="timestampPanel" value="m6" />
            </p:menuButton>

            <div>
                <p:commandButton type="button" value="hack m4" onclick="PrimeFaces.ab({s:'form:m4',p:'form',u:'form',f:'form'})"
                    style="background:orange" />
                <p:commandButton type="button" value="hack m5" onclick="PrimeFaces.ab({s:'form:m5',p:'form',u:'form',f:'form'})"
                    style="background:salmon" />
                <p:commandButton type="button" value="hack m6" onclick="PrimeFaces.ab({s:'form:m6',p:'form',u:'form',f:'form'})" />
            </div>
        </h:panelGroup>


        <h:panelGroup>
            <p:menuButton value="m7 - m8- m9" rendered="false">
                <p:menuitem id="m7" action="#{testMenuitemBean.updateTimestamp}" process="@form" update="timestampPanel"
                    value="m7" />
                <p:menuitem id="m8" disabled="true" action="#{testMenuitemBean.updateTimestamp}" process="@form"
                    update="timestampPanel" value="m8" />
                <p:menuitem id="m9" rendered="false" action="#{testMenuitemBean.updateTimestamp}" process="@form"
                    update="timestampPanel" value="m9" />
            </p:menuButton>

            <div>
                <p:commandButton type="button" value="hack m7" onclick="PrimeFaces.ab({s:'form:m7',p:'form',u:'form',f:'form'})" />
                <p:commandButton type="button" value="hack m8" onclick="PrimeFaces.ab({s:'form:m8',p:'form',u:'form',f:'form'})" />
                <p:commandButton type="button" value="hack m9" onclick="PrimeFaces.ab({s:'form:m9',p:'form',u:'form',f:'form'})" />
            </div>
        </h:panelGroup>


        <h:panelGroup>
            <p:splitButton id="b4" value="m10 - m11 - m12" action="#{testMenuitemBean.updateTimestamp}" process="@form"
                update="timestampPanel">
                <p:menuitem id="m10" action="#{testMenuitemBean.updateTimestamp}" process="@form" update="timestampPanel"
                    value="m10" />
                <p:menuitem id="m11" disabled="true" action="#{testMenuitemBean.updateTimestamp}" process="@form"
                    update="timestampPanel" value="m11" />
                <p:menuitem id="m12" rendered="false" action="#{testMenuitemBean.updateTimestamp}" process="@form"
                    update="timestampPanel" value="m12" />
            </p:splitButton>

            <div>
                <p:commandButton type="button" value="hack b4" onclick="PrimeFaces.ab({s:'form:b4',p:'form',u:'form',f:'form'})" />
                <p:commandButton type="button" value="hack m10" onclick="PrimeFaces.ab({s:'form:m10',p:'form',u:'form',f:'form'})" />
                <p:commandButton type="button" value="hack m11" onclick="PrimeFaces.ab({s:'form:m11',p:'form',u:'form',f:'form'})"
                    style="background:salmon" />
                <p:commandButton type="button" value="hack m12" onclick="PrimeFaces.ab({s:'form:m12',p:'form',u:'form',f:'form'})" />
            </div>
        </h:panelGroup>


        <h:panelGroup>
            <p:splitButton disabled="true" id="b5" value="m13 - m14 - m15" action="#{testMenuitemBean.updateTimestamp}"
                process="@form" update="timestampPanel">
                <p:menuitem id="m13" action="#{testMenuitemBean.updateTimestamp}" process="@form" update="timestampPanel"
                    value="m13" />
                <p:menuitem id="m14" disabled="true" action="#{testMenuitemBean.updateTimestamp}" process="@form"
                    update="timestampPanel" value="m14" />
                <p:menuitem id="m15" rendered="false" action="#{testMenuitemBean.updateTimestamp}" process="@form"
                    update="timestampPanel" value="m15" />
            </p:splitButton>

            <div>
                <p:commandButton type="button" value="hack b5" onclick="PrimeFaces.ab({s:'form:b5',p:'form',u:'form',f:'form'})" />
                <p:commandButton type="button" value="hack m13" onclick="PrimeFaces.ab({s:'form:m13',p:'form',u:'form',f:'form'})"
                    style="background:orange" />
                <p:commandButton type="button" value="hack m14" onclick="PrimeFaces.ab({s:'form:m14',p:'form',u:'form',f:'form'})"
                    style="background:salmon" />
                <p:commandButton type="button" value="hack m15" onclick="PrimeFaces.ab({s:'form:m15',p:'form',u:'form',f:'form'})" />
            </div>
        </h:panelGroup>


        <h:panelGroup>
            <p:splitButton rendered="false" id="b6" value="m16 - m17 - m18" action="#{testMenuitemBean.updateTimestamp}"
                process="@form" update="timestampPanel">
                <p:menuitem id="m16" action="#{testMenuitemBean.updateTimestamp}" process="@form" update="timestampPanel"
                    value="m16" />
                <p:menuitem id="m17" disabled="true" action="#{testMenuitemBean.updateTimestamp}" process="@form"
                    update="timestampPanel" value="m17" />
                <p:menuitem id="m18" rendered="false" action="#{testMenuitemBean.updateTimestamp}" process="@form"
                    update="timestampPanel" value="m18" />
            </p:splitButton>

            <div>
                <p:commandButton type="button" value="hack b6" onclick="PrimeFaces.ab({s:'form:b6',p:'form',u:'form',f:'form'})" />
                <p:commandButton type="button" value="hack m16" onclick="PrimeFaces.ab({s:'form:m16',p:'form',u:'form',f:'form'})" />
                <p:commandButton type="button" value="hack m17" onclick="PrimeFaces.ab({s:'form:m17',p:'form',u:'form',f:'form'})" />
                <p:commandButton type="button" value="hack m18" onclick="PrimeFaces.ab({s:'form:m18',p:'form',u:'form',f:'form'})" />
            </div>
        </h:panelGroup>
    </p:panelGrid>
</h:form>
```
And the following it's the rendering

![test-page-screen](https://user-images.githubusercontent.com/7235289/74587850-e8d4d480-4ff7-11ea-8c2f-30f9625675c3.png)

As you can see, the salmon colored represent the critical bypass, while the orange colored represent the unwanted behavior (the GUI does not allow to "click" those menuitems, but JS code can execute them anyway).
Unrendered components are safe.

To prevent the critical bypass while having no performance impact, we should update the `UIMenuItem.decode()` adding just the `isDisabled()` check:

```java
@Override
public void decode(FacesContext facesContext) {
  if(isDisabled()) {             // <------------   this check
    return;
  }

  Map<String, String> params = facesContext.getExternalContext().getRequestParameterMap();
  String clientId = getClientId(facesContext);

  if (params.containsKey(clientId)) {
    queueEvent(new ActionEvent(this));
  }

  ComponentUtils.decodeBehaviors(facesContext, this);
}
```

To prevent the unwanted behavior while having **no** performance impact (we got a little speedup instead), we should add the `AbstractMenu.processDecodes()` adding just the `isDisabled()` check:

```java
public void processDecodes(FacesContext context) {
    if (context == null) {
        throw new NullPointerException();
    }

    // Skip processing if our rendered flag is false or if we are disabled
    if (!isRendered() || isDisabled()) { // <------- added disabled check
        return;
    }

    pushComponentToEL(context, null);

    // Process all facets and children of this component
    Iterator kids = getFacetsAndChildren();
    while (kids.hasNext()) {
        UIComponent kid = (UIComponent) kids.next();
        kid.processDecodes(context);
    }

    // Process this component itself
    try  {
        decode(context);
    } catch (RuntimeException e) {
        context.renderResponse();
        throw e;
    } finally  {
        popComponentFromEL(context);
    }
}
```
The downsides of this approach are:

- it isn't global
- it has to be applied to **each** component that can be disabled *and* that can contain `UICommands` (i.e. `Tab` component is another one that's impacted)
- it may lead to side-effects

**Comment by melloware** [comment]
I will modify my PR to just:
1. Add disabled check to UIMenuItem
2. Add disabled check to AbstractMenu 

And that should cover us correct with little to no performance impact?

**Comment by mmariotti** [comment]
@melloware ok for 1., it's totally safe.
For 2. we have to do some testing to exclude potential side-effects.
Furthermore, adding 2. only on AbstractMenu is not enough to solve it entirely, other components are affected, like Tab.
Maybe it's better to add an interface like `DisabledAware` to group all these components first.
Also, maybe it's better to open another specific issue?
What do you think about?

**Comment by melloware** [comment]
Yep open another issue.  We will close this one with just the simply fix of #1 checking for disabled in UIMenuItem.
